File: app/src/main/java/fi/kaupunkifillarit/inject/SystemServicesModule.java
package fi.kaupunkifillarit.inject;

import android.app.Application;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.android.gms.common.GoogleApiAvailability;

import dagger.Module;
import dagger.Provides;

@Module
public final class SystemServicesModule {
    private final Application application;

    public SystemServicesModule(Application application) {
        this.application = application;
    }

    @Provides
    @ApplicationScope
    Application provideApplication() {
        return application;
    }

    @Provides
    @ApplicationScope
    SharedPreferences providePreferenceManager() {
        return PreferenceManager.getDefaultSharedPreferences(application);
    }

    @Provides
    @ApplicationScope
    GoogleApiAvailability provideGoogleApiAvailability() {
        return GoogleApiAvailability.getInstance();
    }
}

File: app/src/main/kotlin/fi/kaupunkifillarit/inject/SystemServicesModule.kt
package fi.kaupunkifillarit.inject

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.google.android.gms.common.GoogleApiAvailability
import dagger.Module
import dagger.Provides
import fi.kaupunkifillarit.KaupunkifillaritApplication
import javax.inject.Singleton

@Module
class SystemServicesModule(private val application: KaupunkifillaritApplication) {
    @Provides
    @Singleton
    @ForApplication
    fun provideApplicationContext(): Context {
        return application
    }

    @Provides
    @Singleton
    fun providePreferenceManager(): SharedPreferences {
        return PreferenceManager.getDefaultSharedPreferences(application)
    }

    @Provides
    @Singleton
    fun provideGoogleApiAvailability(): GoogleApiAvailability {
        return GoogleApiAvailability.getInstance()
    }
}
